entity-service/src/main/kotlin/com/egm/stellio/entity/service/EntityOperationService.kt

```diff
@@ -0,0 +1,139 @@
+     */
+    fun create(entities: List<ExpandedEntity>): BatchOperationResult {
+        val (graph, invalidRelationsErrors) = entitiesEntitiesGraphBuilder.build(entities)
+
+        val (naiveBatchResult, entitiesWithCircularDependencies) = createEntitiesWithoutCircularDependencies(graph)
+        val (circularCreateSuccess, circularCreateErrors) = createEntitiesWithCircularDependencies(
+            entitiesWithCircularDependencies.toList()
+        )
+
+        val success = naiveBatchResult.success.plus(circularCreateSuccess)
+        val errors = invalidRelationsErrors.plus(naiveBatchResult.errors).plus(circularCreateErrors)
+
+        return BatchOperationResult(ArrayList(success), ArrayList(errors))
+    }
+
+    private fun createEntitiesWithoutCircularDependencies(graph: Graph<ExpandedEntity, DefaultEdge>): Pair<BatchOperationResult, Set<ExpandedEntity>> {
+        val batchOperationResult = BatchOperationResult(arrayListOf(), arrayListOf())
+        val temporaryGraph = DirectedPseudograph<ExpandedEntity, DefaultEdge>(DefaultEdge::class.java)
+        Graphs.addGraph(temporaryGraph, graph)
+
+        /**
+         * Gets a list of leaves (i.e. with 0 successors).
+         */
+        fun <T, E> Graph<T, E>.getLeaves(): List<T> = vertexSet().filter { outDegreeOf(it) == 0 }
+
+        do {
+            val leaves = temporaryGraph.getLeaves()
+
+            val res = leaves.parallelStream().map {
+                try {
+                    entityService.createEntity(it)
+                    Pair(it.id, null)
+                } catch (e: BadRequestDataException) {
+                    Pair(null, BatchEntityError(it.id, arrayListOf(e.message)))
+                }
+            }
+
+            res.toList().fold(batchOperationResult, { result, (entity, error) ->
+                entity?.let { result.success.add(it) }
+                error?.let { result.errors.add(it) }
+                result
+            })
+
+            leaves.forEach {
+                temporaryGraph.removeVertex(it)
+            }
+        } while (leaves.isNotEmpty())
+
+        return Pair(batchOperationResult, temporaryGraph.vertexSet())
+    }
+
+    /*
+     * Creates given entities into database.
+     * When there are circular dependencies between nodes, the creation should be achieved in 2 steps:
+     * first create a "temp" entity with almost no attributes
+     * then create the relationships on these attributes
+     */
+    private fun createEntitiesWithCircularDependencies(entities: List<ExpandedEntity>): BatchOperationResult {
+        entities.forEach { entity ->
+            createTempEntityInBatch(entity.id, entity.type, entity.contexts)
+        }
+
+        // TODO improve process, if an entity update fails, we should check linked entities to also not delete them
+        return entities.fold(BatchOperationResult(arrayListOf(), arrayListOf()), { (creations, errors), entity ->
+            try {
+                entityService.appendEntityAttributes(
```

Comment:
Please consider removing  isInBatchContext and validEntities parameters from the appendEntityAttributes method if you don't need them anymore.
Since they were added in the purpose of the Batch creation.

Reply:
I believe I already removed all validEntities/isInBatchContext parameters. Notify me if you still see some remaining.